lib/vnet/daemon/service_darwin.m

```diff
@@ -0,0 +1,147 @@
+//go:build vnetdaemon
+// +build vnetdaemon
+
+// Teleport
+// Copyright (C) 2024 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+#include "common_darwin.h"
+#include "protocol_darwin.h"
+#include "service_darwin.h"
+
+#import <Foundation/Foundation.h>
+
+#include <string.h>
+
+@interface VNEDaemonService () <NSXPCListenerDelegate, VNEDaemonProtocol>
+
+@property(nonatomic, strong, readwrite) NSXPCListener *listener;
+@property(nonatomic, readwrite) BOOL started;
+
+@property(nonatomic, readwrite) NSString *socketPath;
+@property(nonatomic, readwrite) NSString *ipv6Prefix;
+@property(nonatomic, readwrite) NSString *dnsAddr;
+@property(nonatomic, readwrite) NSString *homePath;
+@property(nonatomic, readwrite) dispatch_semaphore_t gotVnetConfigSema;
+
+@end
+
+@implementation VNEDaemonService
+
+- (id)initWithBundlePath:(NSString *)bundlePath {
+  self = [super init];
+  if (self) {
+    // Launch daemons must configure their listener with the machServiceName
+    // initializer.
+    _listener = [[NSXPCListener alloc] initWithMachServiceName:DaemonLabel(bundlePath)];
+    _listener.delegate = self;
+
+    _started = NO;
+    _gotVnetConfigSema = dispatch_semaphore_create(0);
+  }
+  return self;
+}
+
+- (void)start {
+  assert(_started == NO);
```

Comment:
Missing "#include <assert.h>".

Should we do explicit errors instead or asserting? Same throughout.

Reply:
Oh thanks. I copied the asserts from the example launch daemon project, I wasn't aware that I have to include something to use them. But it makes sense now that I think about it.

I think the asserts are fine for what this code is. Most of it is expected to be called only once from the main function of a tsh command and adding error handling between Go and Obj-C makes the code so much more obtuse.

Reply:
I worry about the panicky nature of the asserts, coupled with the lack of automated test coverage here, but up to you.

Reply:
Ah, back when I added this code I read a Polish resource about `assert`. It didn't mention panic at all, only that `abort` is used. I didn't realize that `panic` in Go and `abort` in C are pretty much the same. 🙊 All this time I thought this code is just going to print a friendly warning to stdout. 🥴

I'll see if I can use Obj-C exceptions for this and if they behave well with cgo, if not then I'll add explicit errors.

Reply:
Okay, I just removed calls to `assert` completely. In case of `start`/`stop`/`waitForVnetConfig` they were unnecessary since we already check the state of the daemon when calling those methods. In the case of the `listener` implementation of `NSXPCListenerDelegate`, I don't understand why they were there in the first place and the example project I got it from doesn't explain that. I also couldn't find another `NSXPCListenerDelegate` example which would explicitly check those values like that.

Reply:
LGTM.